README.md
# Inspector

<!-- Active status commented out
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
-->

[![Project Tracker](https://img.shields.io/badge/repo%20status-Project%20Tracker-lightgrey)](https://randomserver.xyz/project-tracker.html)
![Platform](https://img.shields.io/badge/platform-Linux-lightgrey)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![CodeFactor](https://www.codefactor.io/repository/github/strangeranger/inspector/badge)](https://www.codefactor.io/repository/github/strangeranger/inspector)

Inspector is a security tool with the purpose of identifying users who have both successfully and unsuccessfully switched to root or another user on Linux based Distributions. It does this by scanning through `/var/log/auth.log` for specific patterns that indicate specific actions/executed commands.

## Getting Started

### Prerequisites

Install the required dependencies using either of the following commands:

- `python3 -m pip install -r requirements.txt` (installs globally)
- `pipenv install -r requirements.txt` (installs locally via pipenv)

### Installing

All you need to do is download the repository. There are no binaries or anything to install.

`git clone https://github.com/StrangeRanger/inspector/`

## Usage

Because Inspector needs to access `/var/log/auth.log`, you'll be required to execute Inspector with root priviledge:

`sudo python3 inspector.py`

## Supported Distributions

The following is a list of all the Linux Distributions that Inspector officially supports and works on:

| Distributions | Distro Versions         |
| ------------- | ----------------------- |
| Ubuntu        | 20.04<br>16.04<br>18.04 |
| Debian        | 10<br>9                 |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.1.2] - 2021-05-29

### Changed

- Renamed `modules` to `pkg`.
- Moved the contents of `globalvar.py` into `__init__.py`.
- Modified docstrings from Numpy to reST.
- Now uses type hinting.
- Slight refactoring.


## [2.1.1] - 2021-03-27

### Changes

- Updated README.md
- Updated Documentation

### Fixes

- Re-definition found for builtin function PYL-W0622 (@deepsourcebot)
- from module import \* used; unable to detect undefined names PYL-W0401 (@deepsourcebot)
- Unused import from wildcard import found PYL-W0614 (@deepsourcebot)
- Consider using in PYL-R1714 (@deepsourcebot)
- Module imported but unused PYL-W0611 (@deepsourcebot)
- Undefined name detected PYL-E0602 (@deepsourcebot)

## [2.1.0] - 2021-02-05

### Added

- Officially supports Ubuntu 20.04
- Relies on `/etc/shells` to know what shells Inspector should keep track of
- Scripts that test compatibility

### Changed

- Documentation changes
- Styled using [black](https://github.com/psf/black)

## [2.0.0] - 2020-07-26

### Added

- If ran without root privilege, users will be notified and the program will exit
- Will present Distro ID and a version number if run on an unsupported distribution/OS

### Changed

- Moved documentation to mkdocs, which is now hosted on readthedocs.org
- Moved the identifying pattern files (now called identifying logs) to mkdocs on readthedocs.org
- Merged what used to be two different branches for different operating systems, into the same branch
  - Code that is required for specific operating systems will be imported based on what operating system program is being run on
- Removed cron scripts till they can be properly implemented
- Updated code to better follow personal and PEP style guide

### Fixed

- A few small bugs

[unreleased]: https://github.com/StrangeRanger/inspector/compare/v2.1.2...HEAD
[2.1.2]: https://github.com/StrangeRanger/inspector/releases/tag/v2.1.2
[2.1.1]: https://github.com/StrangeRanger/inspector/releases/tag/v2.1.1
[2.1.0]: https://github.com/StrangeRanger/inspector/releases/tag/v2.1.0
[2.0.0]: https://github.com/StrangeRanger/inspector/releases/tag/v2.0.0


pkg/distro_specific.py
"""
Code that varies depending on the Linux Distribution inspector is run on.
"""
####[ Imports ]#########################################################################


from datetime import datetime
from pkg import (
    CYAN,
    DEFCLR,
    today,
    start_date,
    this_year,
    last_year,
    days,
    daysv2,
)


####[ Class creation ]##################################################################


class DateError(Exception):
    """Raise when log date is incorrect.

    See comment under the line containing 'date_str = " ".join(fields[0:2]) + " "'
    more information.
    """


####[ Functions ]#######################################################################


def debian10_ubuntu20(file: str) -> None:
    """Look through 'file' and find specific logs that can be used to identify actions
    performed by a user.

    :param file: Name of the file to be inspected.
    :note:       Here is a list of Linux Distributions this function is used on:
                     - Ubuntu 20.04
                     - Debian 10
    """
    su_bin = ["COMMAND=/bin/su", "COMMAND=/usr/bin/su"]
    shell_bin = []
    shells = open("/etc/shells", "r")

    for line in shells:
        shell_bin.append("COMMAND={}".format(line.strip()))

    for line in file:
        fields = line.split()
        date_str = " ".join(fields[0:2]) + " "
        ## Makes sure that the log date is correct; if current date is January 01
        ## 2020 and looking a line in log with date Dec 31 that was logged in
        ## 2019,the date would = Dec 31 2020. Lines below prevent this.
        try:
            date = datetime.strptime(date_str + str(this_year), "%b %d %Y").date()
            if date > today:
                raise DateError
        except DateError:
            date = datetime.strptime(date_str + str(last_year), "%b %d %Y").date()
        ## Will skip any abnormal/non-regular text in '/var/log/auth.log' that could
        ## produce an Error, and then prints out a message telling the user to
        ## checkout the line in the file.
        except ValueError:
            print(
                "{}There was an abnormality in /var/log/auth.log: {}{}\n".format(
                    CYAN, line, DEFCLR
                )
            )
            continue

        if date < start_date:
            # Too old for interest
            continue
        ## "user : TTY=tty/1 ; PWD=/home/user ; USER=root ; COMMAND=/bin/su"
        if fields[4] == "sudo:":
            user = fields[5]
            # Successful.
            conditions = (
                user != "root"
                and (
                    fields[8] not in ("incorrect", "NOT") if len(fields) >= 9 else None
                )
                and fields[-4] == "USER=root"
                and fields[-2] in su_bin
            )
            # Unsuccessful.
            conditions2 = (
                user != "root"
                and (fields[8] == "incorrect" if len(fields) >= 9 else None)
                and fields[-4] == "USER=root"
                and fields[18] in su_bin
            )
            conditions3 = fields[-3] == "USER=root" and fields[-1] in su_bin
            # `sudo -i` and `sudo bash`
            # Purposefully separated from conditions3,
            # else when previous commands are used, the number of times a user
            # attempts to log into an account becomes twice as many as actual;
            # without writing a paragraph, this is due to what text is being
            # looked for, and what is produced when the commands above are (in
            # this comment) are executed
            conditions35 = fields[-3] == "USER=root" and fields[-1] in shell_bin

            ## "..."; identifies users who are not in the sudoers file and tried
            ## to execute a command with root privilege.
            if user != "root" and (
                fields[8] == "NOT"
                and fields[10] == "sudoers"
                and fields[16] == "USER=root"
                and fields[18].startswith("COMMAND=")
                if len(fields) >= 9
                else None
            ):
                days[date]["~" + user] += 1
            ## "..."; identifies users who successfully became root using `sudo
            ## bash` or `sudo -i`
            if (
                user != "root"
                and (fields[8] != "incorrect" if len(fields) >= 9 else None)
                and conditions35
            ):
                # A.2. The defaultdict key becomes the date and its value, which
                # is the counter, is the user, which gains a plus 1 in the counter
                days[date]["+" + user] += 1
            ## "..."; identifies users who unsuccessfully became root using `sudo
            ## bash` or `sudo -i`
            elif (
                user != "root"
                and (fields[8] == "incorrect" if len(fields) >= 9 else None)
                and conditions35
            ):
                days[date]["*" + user] += int(fields[7])  # A.2.
            ## "..."; identifies users who unsuccessfully became root using `sudo su`
            elif (
                user != "root"
                and (fields[8] == "incorrect" if len(fields) >= 9 else None)
                and conditions3
            ):
                days[date]["*" + user] += int(fields[7])  # A.2.
            ## "..."; identifies users who unsuccessfully became root using `sudo
            ## su root`
            elif conditions2 and fields[-1] == "root":
                days[date]["*" + user] += int(fields[7])  # A.2.
            ## "..."; identifies users who unsuccessfully switched users using
            ## `sudo su <username>`
            elif conditions2 and fields[-1] != "root":
                victim = fields[19]
                daysv2[date]["/" + user][victim] += int(fields[7])  # B.2.

        if fields[4].startswith("su:"):
            # (to root) <username>
            conditions4 = fields[-4] == "root)" and fields[-3] != "root"
            # (to <victim_username>) <username>
            conditions5 = fields[-4] != "root)" and fields[-3] != "root"

            ## "(to root) <username> on pts/<n>"; identifies users who've successfully
            ## became root using `su`, `su root`, `sudo su`, or `sudo su root`
            if " ".join(fields[5:7]) == "(to root)" and conditions4:
                user = fields[-3]
                days[date]["+" + user] += 1  # A.2.
            ## "FAILED SU (to root) <username> on pts/<n>"; identifies users who've
            ## unsuccessfully became root using `su` and/or `su root`
            elif " ".join(fields[5:7]) == "FAILED SU" and conditions4:
                user = fields[-3]
                days[date]["*" + user] += 1  # A.2.
            ## "(to <victim_username>) <username> on pts/<n>"; identifies users
            ## who've successfully switched users using `su <victim_username>`
            ## or `sudo su <victim_username>`
            elif fields[5] == "(to" and conditions5:
                user = fields[-3]
                victim = fields[-4].replace(")", "")
                daysv2[date]["-" + user][victim] += 1  # B.2.
            ## "FAILED SU (to <victim_username>) <username> on pts/<n>"; identifies
            ## users who've unsuccessfully switched users using `su <victim_username>`
            elif " ".join(fields[5:7]) == "FAILED SU" and conditions5:
                user = fields[-3]
                victim = fields[-4].replace(")", "")
                daysv2[date]["/" + user][victim] += 1  # B.2.


def debian9_ubuntu16(file):
    """Look through 'file' and find specific logs that can be used to identify actions
    performed by a user.

    :param file: Name of the file to be inspected.
    :note:       Here is a list of Linux Distributions this function is used on:
                     - Ubuntu 16.04
                     - Ubuntu 18.04
                     - Debian 9
    """
    su_bin = ["COMMAND=/bin/su", "COMMAND=/usr/bin/su"]
    shell_bin = ["COMMAND=/bin/su", "COMMAND=/usr/bin/su"]
    shells = open("/etc/shells", "r")

    for line in shells:
        shell_bin.append("COMMAND={}".format(line.strip()))

    for line in file:
        fields = line.split()
        date_str = " ".join(fields[0:2]) + " "
        ## Makes sure that the log date is correct; if current date is January 01
        ## 2020 and looking a line in log with date Dec 31 that was logged in
        ## 2019,the date would = Dec 31 2020. Lines below prevent this.
        try:
            date = datetime.strptime(date_str + str(this_year), "%b %d %Y").date()
            if date > today:
                raise DateError
        except DateError:
            date = datetime.strptime(date_str + str(last_year), "%b %d %Y").date()
        ## Will skip any abnormal/non-regular text in '/var/log/auth.log' that could
        ## produce an Error, and then prints out a message telling the user to
        ## checkout the line in the file.
        except ValueError:
            print(
                "{}There was an abnormality in /var/log/auth.log: {}{}\n".format(
                    CYAN, line, DEFCLR
                )
            )
            continue

        if date < start_date:
            # Too old for interest
            continue
        ## "user : TTY=tty/1 ; PWD=/home/user ; USER=root ; COMMAND=/bin/su"
        if fields[4] == "sudo:":
            user = fields[5]
            ## Successful
            conditions = (
                user != "root"
                and (
                    fields[8] not in ("incorrect", "NOT") if len(fields) >= 9 else None
                )
                and fields[-4] == "USER=root"
                and fields[-2] in su_bin
            )
            ## Unsuccessful
            conditions2 = (
                user != "root"
                and (fields[8] == "incorrect" if len(fields) >= 9 else None)
                and fields[-4] == "USER=root"
                and fields[-2] in su_bin
            )
            # `sudo su`...
            conditions3 = fields[-3] == "USER=root" and fields[-1] in shell_bin

            ## "..."; identifies users who are not in the sudoers file and tried to
            ## execute a command with root privilege
            if user != "root" and (
                fields[8] == "NOT"
                and fields[10] == "sudoers"
                and fields[16] == "USER=root"
                and fields[18].startswith("COMMAND=")
                if len(fields) >= 9
                else None
            ):
                days[date]["~" + user] += 1
            ## "..."; identifies users who successfully became root using `sudo su`
            if (
                user != "root"
                and (fields[8] != "incorrect" if len(fields) >= 9 else None)
                and conditions3
            ):
                # A.2. The defaultdict key becomes the date and its value, which
                # is the counter, is the user, which gains a plus 1 in the counter
                days[date]["+" + user] += 1
            ## "..."; identifies users who unsuccessfully became root using `sudo su`
            elif (
                user != "root"
                and (fields[8] == "incorrect" if len(fields) >= 9 else None)
                and conditions3
            ):
                days[date]["*" + user] += int(fields[7])  # A.2.
            ## "..."; identifies users who successfully became root using `sudo
            ## su root`
            elif conditions and fields[-1] == "root":
                days[date]["+" + user] += 1  # A.2.
            ## "..."; identifies users who unsuccessfully became root using
            ## `sudo su root`
            elif conditions2 and fields[-1] == "root":
                days[date]["*" + user] += int(fields[7])  # A.2.
            ## "..."; identifies users who successfully switched users using
            ## `sudo su <username>`
            elif conditions and fields[-1] != "root":
                victim = fields[14]
                daysv2[date]["-" + user][victim] += 1  # B.2.
            ## "..."; identifies users who unsuccessfully switched users using
            ## `sudo su <username>`
            elif conditions2 and fields[-1] != "root":
                victim = fields[19]
                daysv2[date]["/" + user][victim] += int(fields[7])  # B.2.

        if fields[4].startswith("su["):
            # root by <username>
            conditions4 = fields[-3] == "root" and fields[-1] != "root"
            # <username> by <username>
            conditions5 = "root" not in (fields[-3], fields[-1])

            ## "Successful su for root by <username>"; identifies users who've
            ## successfully became root using `su` and/or `su root`
            if fields[5] == "Successful" and conditions4:
                user = fields[-1]
                days[date]["+" + user] += 1  # A.2.
            ## "FAILED su for root by <username>"; identifies users who've
            ## unsuccessfully became root using `su` and/or `su root`
            elif fields[5] == "FAILED" and conditions4:
                user = fields[-1]
                days[date]["*" + user] += 1  # A.2.
            ## "Successful su for <username> by <username>"; identifies users
            ## who've successfully switched users using `su <username>`
            elif fields[5] == "Successful" and conditions5:
                user = fields[-1]
                victim = fields[-3]
                daysv2[date]["-" + user][victim] += 1  # B.2.
            ## "FAILED su for <username> by <username>"; identifies users who've
            ## unsuccessfully switched users using `su <username>`
            elif fields[5] == "FAILED" and conditions5:
                user = fields[-1]
                victim = fields[-3]
                daysv2[date]["/" + user][victim] += 1  # B.2.
